bpf/bpf_host.c

```diff
@@ -170,8 +170,7 @@ resolve_srcid_ipv6(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
 	if (from_host)
 		src_id = srcid_from_ipcache;
 	else if (src_id == WORLD_ID &&
-		 identity_from_ipcache_ok() &&
-		 !identity_is_reserved(srcid_from_ipcache))
+		 identity_from_ipcache_ok())
```

Comment:
Is it possible for NodePort traffic to hit this condition? If yes, and if the source is masqueraded, then the IP will look like a host IP, regardless of whether the traffic came from the host. That would be a security hole. I think that's the real reason why there's a reserved check here.

Reply:
cc @pchaigno , hopefully the above query is just noise but I would like to make sure given the implications of the change.

Reply:
Hm, I hesitated on blocking this on your review. Maybe I should have :smile: 

> Is it possible for NodePort traffic to hit this condition?

I had the same concern. My understanding however is that this is for traffic hitting tc-ingress of the native devices. So kube-proxy didn't have a chance to interfere yet. And KPR didn't run yet when we execute this bit of code (that happens after in `CILIUM_CALL_IPV{4,6}_FROM_NETDEV`).

Reply:
> Is it possible for NodePort traffic to hit this condition? If yes, and if the source is masqueraded, then the IP will look like a host IP, regardless of whether the traffic came from the host. That would be a security hole. I think that's the real reason why there's a reserved check here.

On a related note, do we have a CI test which asserts future changes don't expose ourselves to this security hole?

Reply:
bpf_host is also attached to `cilium_host`:

```
# bpftool net
xdp:

tc:
cilium_net(2) clsact/ingress bpf_host_cilium_net.o:[to-host] id 1670
cilium_host(3) clsact/ingress bpf_host.o:[to-host] id 1656
cilium_host(3) clsact/egress bpf_host.o:[from-host] id 1663
...
```

Isn't it possible for a packet to arrive at some native device, Cilium passes it up the stack towards kube-proxy for nodeport handling, then after DNAT+SNAT the packet is routed towards `cilium_host` (endpoint-routes=false) -- then won't this code run?

Reply:
At least in the Jenkins suite, we have the test where a 3rd node without Cilium is installed that can be used to simulate this scenario. If we don't want Jenkins / Vagrant tests anymore, then it seems plausible that we could have a GH Actions job with a kind cluster and initiate these nodeport requests from outside of the cluster.

Reply:
> bpf_host is also attached to `cilium_host`:

The above code is for the `!from_host` case.

> I think the tricky part here is that we would be configuring a nodeport, sending traffic from outside the node towards that nodeport, applying a security policy that drops the traffic, then the test would validate that the traffic is in fact dropped.

I think we already have at least one test covering that. See the `with L7 policy` test in the NodePort tests of services.go. That one is currently flaky but you get the idea. And it might not be the only NodePort test with policy enforcement.

Reply:
> bpf_host is also attached to `cilium_host`:

The above code is for the `!from_host` case.

> I think the tricky part here is that we would be configuring a nodeport, sending traffic from outside the node towards that nodeport, applying a security policy that drops the traffic, then the test would validate that the traffic is in fact dropped.

I think we already have at least one test covering that. See the `with L7 policy` test in the NodePort tests of services.go. That one is currently flaky but you get the idea. And it might not be the only NodePort test with policy enforcement.

Reply:
>Isn't it possible for a packet to arrive at some native device, Cilium passes it up the stack towards kube-proxy for nodeport handling, then after DNAT+SNAT the packet is routed towards cilium_host (endpoint-routes=false) -- then won't this code run?

Hmm, it could... 🤔  We store the original source identity in ctx (see `nodeport_lb6`) when nodeport (kpr) is enabled. If we saved the identity unconditionally, it could help the cilium_host case. Although I'm not sure if the scratch pad is reset when packets are passed through stack, so we'll probably lose access to the encoded identity information. We only store identities in packet mark for local pods at the moment.

Reply:
> > bpf_host is also attached to cilium_host:
> 
> The above code is for the !from_host case.

:ok_hand: Awesome, then any time that iptables is involved then traffic is routed to cilium_host, it will not hit this line. Great.

One more case: XDP -> TC on receive into the node. I presume that we reliably either don't SNAT or we hand identity context between XDP->TC for native device?

Reply:
> Although I'm not sure if the scratch pad is reset when packets are passed through stack, so we'll probably lose access to the encoded identity information. 

Yeah this is a common difficulty while routing traffic through the stack. We may have more control over this in the "bpf host routing" case.

Reply:
> One more case: XDP -> TC on receive into the node. I presume that we reliably either don't SNAT or we hand identity context between XDP->TC for native device?

That's reaching outside of my knowledge. You'd have to find someone familiar with that or dig the answer yourself.

Reply:
cc @borkmann / @cilium/sig-datapath  would you happen to know about the above question or be able to point towards someone with that context? I _think_ the first statement is true, not so sure about the second statement.
- Don't have a path where Cilium XDP applies SNAT then forwards up the stack, and
- If forwarding up the stack, we ensure that XDP passes the source identity up the stack

If these both hold then I think we're good to proceed with this PR.

Reply:
I'm digging the code and couldn't find a path where Cilium XDP applies SNAT then forwards the packet up the stack. But we should wait for the datapath team's comment.


In XDP Cilium calls nodeport_lb[4,6]

1. Lookup service and select a backend for the svc and do DNAT. https://github.com/cilium/cilium/blob/9f35b129dfb746f2f4a12d614cfefe99088acb2b/bpf/lib/nodeport.h#L892
2. If the backend is local, writes XFER_PKT_NO_SVC and return CTX_ACT_OK. https://github.com/cilium/cilium/blob/9f35b129dfb746f2f4a12d614cfefe99088acb2b/bpf/lib/nodeport.h#L962
   - If the backend is remote, then SNAT and redirect.(lb_mode = snat) Not move up to stack in this case. https://github.com/cilium/cilium/blob/9f35b129dfb746f2f4a12d614cfefe99088acb2b/bpf/lib/nodeport.h#L995
3. Move xfer to data_meta and up to stack
4. In TC, get xfer and set nodeport skip flag. https://github.com/cilium/cilium/blob/0670fd2ace1136a1143a384b39912f8153d7a063/bpf/bpf_host.c#L1056
5. Call resolve_srcid_ipv[4,6] and resolve an identity from ipcache. The source IP is a client IP at this point.

There's a special loopback case in which the origin endpoint has sent to a service which is being translated back to the source, and the source IP will be IPV4_LOOPBACK.
https://github.com/cilium/cilium/blob/ff05a619168d7c4b25cc135e612da1f740691e8c/bpf/lib/lb.h#L1705

But this case is nothing to do with N/S traffic. It's not the case.
https://github.com/cilium/cilium/blob/29479334c197748fa1ad1bcb990a8d1e8a7933c3/bpf/bpf_lxc.c#L1748

Reply:
In XDP to TC punt case, Cilium does SNAT in XDP and moves up the stack, but it encaps and redirects in cil_from_netdev. So it doesn't call resolve_srcid_ipv[4,6]

https://github.com/cilium/cilium/blob/9f35b129dfb746f2f4a12d614cfefe99088acb2b/bpf/lib/nodeport.h#L827
https://github.com/cilium/cilium/blob/0670fd2ace1136a1143a384b39912f8153d7a063/bpf/bpf_host.c#L1065

Reply:
> If these both hold then I think we're good to proceed with this PR.

If the first is true then why does the second matter? If the packet wasn't SNATed when it reaches the tv hook, then we can trust the ipcache anyway.

Reply:
@joestringer As far as I looked into the datapath code and tested it, Cilium doesn't apply SNAT for the inbound packets to local backends. So we can trust the ipcache here because the source is a client IP. What do you think?

Reply:
> when XDP "passes up the stack" and then some iptables rule munges the packets before the packet is delivered to TC-BPF for pod ingress processing

Is there any chance that iptables mungles the packets between XDP and TC? Did I miss something?

BTW, HOST_ID resolved from the ipcaches is ignored for IPV4 case, but no corresponding check for IPv6. Is there any reason for that?
https://github.com/cilium/cilium/blob/8ea979cfab6ac1005efc591a1f0378333bba6848/bpf/bpf_host.c#L408-L419

Reply:
> > when XDP "passes up the stack" and then some iptables rule munges the packets before the packet is delivered to TC-BPF for pod ingress processing
> 
> Is there any chance that iptables mangles the packets between XDP and TC? Did I miss something?

If we're speaking about XDP and tc on the same device, then no, there are no netfilter hooks between XDP and tc-bpf ingress.

Reply:
I see, now we might bifurcate a bit..

One case I had in mind is with XDP at the native device and TC-BPF ingress at the veth device, which could be the case with endpoint-routes mode.

The latter case I'm not sure about would be endpoint-routes disabled, do we do TC-BPF on the native device for that case these days? Previously or perhaps on earlier kernels I was under the impression that would occur on another device post-routing, maybe cilium_host device.

Reply:
> BTW, HOST_ID resolved from the ipcache is ignored for IPV4 case, but no corresponding check for IPv6. Is there any reason for that? 

I think this is the same question as I asked in this thread?

https://github.com/cilium/cilium/pull/23091#discussion_r1088128446

Reply:
> One case I had in mind is with XDP at the native device and TC-BPF ingress at the veth device, which could be the case with endpoint-routes mode.

The identity resolution will be handled on the bpf_lxc side in this case, and the HOST_ID resolved from ipcache will be ignored here.

https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1581-L1592
https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1926-L1937

> The latter case I'm not sure about would be endpoint-routes disabled, do we do TC-BPF on the native device for that case these days? Previously or perhaps on earlier kernels I was under the impression that would occur on another device post-routing, maybe cilium_host device.

I assume that this case is under BPF HOST routing=false and Endpoint routing=false. This patch doesn't touch from_host path as Paul mentioned https://github.com/cilium/cilium/pull/23091/files#discussion_r1088422353

TC BPF ingress on netdev
 -> resolve_srcid_ipv[4,6](ctx, identity, from_host=false)
 -> pass to stack
 -> TC BPF ingress on cilium_host
 -> resolve_srcid_ipv[4,6](ctx, identity, from_host=true)
 -> ipv[4, 6]_local_delivery
 -> tail_call_dynamic
 -> redirect_peer
 -> Pod ns

Reply:
> The latter case I'm not sure about would be endpoint-routes disabled, do we do TC-BPF on the native device for that case these days? Previously or perhaps on earlier kernels I was under the impression that would occur on another device post-routing, maybe cilium_host device.

With endpoint-routes disabled, we attach tc BPF (bpf_host) to the native device.

Reply:
> The identity resolution will be handled on the bpf_lxc side in this case, and the HOST_ID resolved from ipcache will be ignored here.

:+1: makes sense, thanks for checking!

> With endpoint-routes disabled, we attach tc BPF (bpf_host) to the native device.

:+1: Right yep and there's no iptables hook between. That resolves my concerns.

Reply:
> Why don't we add the HOST_ID check to resolve_srcid_ipv6 as well? 

I think we should add that for the IPv6 case too.

Reply:
Revisiting one more time, Cilium v1.12.7, running inside the Cilium container:

```
# bpftool net
xdp:

tc:
cilium_net(2) clsact/ingress bpf_host_cilium_net.o:[to-host] id 11473
cilium_host(3) clsact/ingress bpf_host.o:[to-host] id 11457
cilium_host(3) clsact/egress bpf_host.o:[from-host] id 11465
cilium_vxlan(4) clsact/ingress bpf_overlay.o:[from-overlay] id 11442
cilium_vxlan(4) clsact/egress bpf_overlay.o:[to-overlay] id 11449
lxc76d5c04bf2ff(8) clsact/ingress bpf_lxc.o:[from-container] id 11541
lxc0443326c1546(10) clsact/ingress bpf_lxc.o:[from-container] id 11552
lxcd6a0dcb63216(12) clsact/ingress bpf_lxc.o:[from-container] id 11532
lxc2d9f8e8642ce(14) clsact/ingress bpf_lxc.o:[from-container] id 11535
lxc21b81bb5a3ab(16) clsact/ingress bpf_lxc.o:[from-container] id 11547
lxcd4cb3575c022(18) clsact/ingress bpf_lxc.o:[from-container] id 11546
lxcd398a5dcaa1d(22) clsact/ingress bpf_lxc.o:[from-container] id 11558
lxc4e2dc37164c3(26) clsact/ingress bpf_lxc.o:[from-container] id 11538
lxc_health(28) clsact/ingress bpf_lxc.o:[from-container] id 11587

flow_dissector:

```

So we see that the `cilium_host` device has the `bpf_host.o` program attached, both for `to-host` and `from-host`. I guess here we're more concerned about `from-host`. Tracing through the from-host section to `handle_netdev()` -> `do_netdev()` .... we may end up `inherit_identity_from_host()`, but then subsequently we follow through to `resolve_srcid_ipv4()` or `resolve_srcid_ipv6()`. So then yeah we need to make absolutely sure that if any traffic hit the stack and did SNAT to a host IP then was directed towards that device, we don't treat such traffic as "from the host".

I was under the impression that we avoided this with the `identity_is_reserved()` check that is changed in this PR, but that's not the case. Only in the IPv4 case we avoid this with the `if (*sec_label != HOST_ID)` check that @ysksuzuki highlighted above. So we've had the bug in the IPv6 path all along for a while.

This also means that this PR is innocent in this regard, it's not impacting this constraint. But yeah we need to follow up on the IPv6 side.